### ISSUE TITLE
为什么logits要在softmax前变为fp32

### ISSUE THREAD

**Issue by llltttppp** [QwenLM/Qwen2#570]
hidden_states = outputs[0]
        logits = self.lm_head(hidden_states)
        logits = logits.float()

modeling中为什么要把logits转为float呢？这显著增加了显存开销！！！！https://github.com/huggingface/transformers/blob/v4.41.2/src/transformers/models/qwen2/modeling_qwen2.py#L1163

**Comment by jklj077** [comment]
Hi,

For inference, I don't think it will increase memory substantially. The logits is of size [batch_size, 1, num_embeddings], which is less than 0.6MiB per sample. 

For training, you will want to keep the input of softmax in fp32, which is a good pratice to maintain stability.

**Comment by Chandler-Bing** [comment]
> Hi,
> 
> For inference, I don't think it will increase memory substantially. The logits is of size [batch_size, 1, num_embeddings], which is less than 0.6MiB per sample.
> 
> For training, you will want to keep the input of softmax in fp32, which is a good pratice to maintain stability.
hi, @jklj077  
For inference, logits size is [batch_size, seq_length, vocab_size], actually when input seq_len is pretty big like 128k, it do substantially increase memory.

**Comment by jklj077** [comment]
Hi, @Chandler-Bing , thanks for the reminder. 

After checking `transformers`, I agree with you (for the first decoding step).

(For future reference) In the context of inference of causal LMs, the process typically comprises two main phases: prefilling and decoding. The prefilling phase calculates and caches key (K) and value (V) matrices for self-attention, while the output logits are not required. Conversely, the decoding phase involves generating logits and sampling tokens sequentially, utilizing the cached K and V matrices, without needing the full sequence's logits.

An efficient approach would involve passing the user input, minus the last token, through the model to generate the cache, without the need of the logits. Subsequently, the last user input token and cache are iteratively fed into the model to obtain the logits and subsequent tokens, appending each new token to the input. This way, the logits is ensured to have dimensions [batch_size, 1, num_embeddings].

However, that's not the case in the `transformers` implementation. As per [here](https://github.com/huggingface/transformers/blob/b7ee1e80b912c6cdd93b54dd77af061fde151d28/src/transformers/generation/utils.py#L2646-L2663) and [here](https://github.com/huggingface/transformers/blob/b7ee1e80b912c6cdd93b54dd77af061fde151d28/src/transformers/models/qwen2/modeling_qwen2.py#L1274-L1303), the entire user input sequence is processed to just obtain the logits of the last token during the initial step, resulting in logits of size [batch_size, user_input_length, num_embeddings]. Thereafter, the last token and cache are used for subsequent steps, yielding logits of size [batch_size, 1, num_embeddings].

While this approach may not be the most optimized, it's not unexpected for `transformers`, which is a more general-purpose library for LLMs. For highly optimized inference, specialized frameworks like vLLM are recommended.

**Comment by Chandler-Bing** [comment]
thanks for answering,
I agree with u @jklj077 ,transformers implementation is more universal, according to this  [Improving memory efficiency](https://github.com/huggingface/transformers/issues/30860), I think they're working on it.
